# Data Loading

Load and examine the initial structure of the datasets required for this analysis.

- **orders.csv:** order-level KPIs (revenue, orders, AOV) and time trends  
- **purchases.csv:** item-level product details (one row per item)
- **sessions.csv:** provides device/viewport/geo context by user session
- **checkout_dropoff.csv:** session-level checkout outcomes
- **events_funnel.csv:** event-level tracking for funnel steps

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Show all columns
pd.set_option("display.max_columns", None)

#### Orders Data: Load `orders.csv` and parse relevant columns as dates to facilitate time-based analysis.

In [2]:
df_orders = pd.read_csv('data/orders.csv')

In [3]:
df_orders.head()

transaction  id                       session_key  \
0  3c393f91-9dac-4d1b-bc9c-df7aaa77ae6c  1289644372.1769538534-1769538533   
1  8c216563-36e5-48a8-8d0e-27d675e01d08  1833988998.1768880128-1769147505   
2  65f1359a-5175-4722-b4f5-4a7e9d678b42   531065886.1769545136-1769545135   
3  d59d8c94-1761-4cd2-8628-666c7d2024cf  1833988998.1768880128-1769147505   
4  8ff144ec-7231-414a-909e-94452fdfc7c7  1712912575.1769517746-1769543575   

                      purchase_ts currency  revenue  units  item_lines  \
0  2026-01-27 18:31:37.192183 UTC      USD     8.99      1           1   
1  2026-01-23 06:45:24.912877 UTC      USD     8.99      1           1   
2  2026-01-27 20:19:19.650985 UTC      USD     8.99      1           1   
3  2026-01-23 06:31:45.199547 UTC      USD     9.75      1           1   
4  2026-01-27 20:00:04.168122 UTC      USD     9.75      1           1   

   distinct_items  
0               1  
1               1  
2               1  
3               1  
4               1

#### Display the first five individual Product Data rows in `purchases.csv`.

In [4]:
df_products = pd.read_csv('data/purchases.csv')

In [5]:
df_products.head()

ts  user_pseudo_id  \
0  2026-01-27 22:13:42.050580 UTC    2.013145e+09   
1  2026-01-27 20:20:52.912429 UTC    7.981743e+08   
2  2026-01-27 20:56:00.936816 UTC    6.992121e+08   
3  2026-01-27 18:29:41.825637 UTC    1.289644e+09   
4  2026-01-27 21:25:54.981308 UTC    1.262121e+09   

                        session_key                        transaction id  \
0  2013144899.1769517745-1769551745  3016e6d2-5c84-4916-bb19-61bd73b8b43d   
1   798174309.1769545220-1769545220  4705c305-b905-44dd-8980-0ee78b1cb743   
2   699212136.1769547295-1769547294  d4b5ce36-4972-481d-8f3c-da492a223d81   
3  1289644372.1769538534-1769538533  d6e4db15-8363-491e-be75-f4e3b2b57960   
4  1262120529.1769549106-1769549106  a8b410c3-3789-4f0a-aa6f-9e5ff6a5098f   

   order_value currency  item_id   item_name  price  quantity  item_revenue  
0         16.0      USD       11  Product 11   16.0         1          16.0  
1         39.0      USD       15  Product 15   39.0         1          39.0  
2         18.5      USD        2   Product 2   18.5         1          18.5  
3         22.0      USD       10  Product 10   22.0         1          22.0  
4         29.0      USD       12  Product 12   29.0         1          29.0

#### Display the first five Session data rows of `sessions.csv`.

In [6]:
df_sessions = pd.read_csv('data/sessions.csv')

In [7]:
df_sessions.head(10)

session_key  user_pseudo_id  \
0  1766310541.1769539027-1769539027    1.766311e+09   
1  1675157078.1769508380-1769508380    1.675157e+09   
2   954119712.1769507864-1769507864    9.541197e+08   
3  1124376317.1769540077-1769540076    1.124376e+09   
4  1309400224.1769507695-1769507695    1.309400e+09   
5   133880033.1769516863-1769516863    1.338800e+08   
6  1962672476.1769514774-1769514773    1.962672e+09   
7  1845904191.1769208898-1769208898    1.845904e+09   
8  1562759019.1769170814-1769170814    1.562759e+09   
9  1300802075.1769507843-1769507843    1.300802e+09   

                 session_start_ts landing_page_path session_viewport_bucket  \
0  2026-01-27 18:37:07.703452 UTC               NaN                     NaN   
1  2026-01-27 10:06:20.526923 UTC               NaN                     NaN   
2  2026-01-27 09:57:44.338723 UTC               NaN                     NaN   
3  2026-01-27 18:54:37.022376 UTC               NaN                     NaN   
4  2026-01-27 09:54:55.567927 UTC               NaN                     NaN   
5  2026-01-27 12:27:44.381649 UTC               NaN                     NaN   
6  2026-01-27 11:52:53.726970 UTC                 /                 360-389   
7  2026-01-23 22:54:58.676430 UTC                 /                 360-389   
8  2026-01-23 12:20:14.817100 UTC                 /                 360-389   
9  2026-01-27 09:57:23.620639 UTC                 /                 360-389   

  session_viewport_profile session_viewport_group country   region  \
0                      NaN                    NaN  Canada  Ontario   
1                      NaN                    NaN  Canada  Ontario   
2                      NaN                    NaN  Canada  Ontario   
3                      NaN                    NaN  Canada  Ontario   
4                      NaN                    NaN  Canada  Ontario   
5                      NaN                    NaN  Canada  Ontario   
6                      NaN                    NaN  Canada  Ontario   
7                      NaN                    NaN  Canada  Ontario   
8                      NaN                    NaN  Canada  Ontario   
9                      NaN                    NaN  Canada  Ontario   

  device_category         os browser  
0         desktop  Macintosh  Chrome  
1          mobile        iOS  Safari  
2          mobile        iOS  Safari  
3         desktop  Macintosh  Chrome  
4          mobile        iOS  Safari  
5         desktop  Macintosh  Chrome  
6         desktop  Macintosh  Chrome  
7         desktop  Macintosh  Chrome  
8         desktop  Macintosh  Chrome  
9          mobile        iOS  Safari

#### Display the first five Checkout data rows in `checkout_dropoff.csv`.

In [8]:
df_checkout = pd.read_csv('data/checkout_dropoff.csv')

In [9]:
df_checkout.head()

session_key               begin_checkout_ts  \
0    68452117.1769553712-1769553711  2026-01-27 22:44:20.992607 UTC   
1   591413466.1769553485-1769553484  2026-01-27 22:38:24.240412 UTC   
2   618581068.1769553392-1769553391  2026-01-27 22:36:37.685767 UTC   
3    43482501.1769517775-1769552743  2026-01-27 22:26:50.381091 UTC   
4  2013144899.1769517745-1769551745  2026-01-27 22:10:44.128115 UTC   

                      purchase_ts                        transaction_id  \
0                             NaN                                   NaN   
1  2026-01-27 22:38:24.430573 UTC  4753c1b2-75ce-4c57-ada0-775eaa3fa757   
2                             NaN                                   NaN   
3                             NaN                                   NaN   
4  2026-01-27 22:10:44.174588 UTC  c51fca48-0ede-4edc-9692-538a3a6a980d   

   purchase_value  is_dropoff  begin_checkout_items_len  
0             NaN           1                         0  
1            39.0           0                         0  
2             NaN           1                         0  
3             NaN           1                         0  
4            43.5           0                         0

#### Display the first five Funnel event data rows in `events_funnel.csv`.

In [10]:
df_funnel = pd.read_csv('data/events_funnel.csv')

In [11]:
df_funnel.head()

ts      event_name  user_pseudo_id  \
0  2026-01-27 22:44:20.992607 UTC  begin_checkout    6.845212e+07   
1  2026-01-27 22:44:19.077762 UTC     select_item    6.845212e+07   
2  2026-01-27 22:44:19.077762 UTC       view_item    6.845212e+07   
3  2026-01-27 22:44:19.077762 UTC       view_item    6.845212e+07   
4  2026-01-27 22:44:02.178645 UTC     select_item    6.845212e+07   

                      session_key   page_path transaction_id  value currency  \
0  68452117.1769553712-1769553711   /checkout            NaN    NaN      NaN   
1  68452117.1769553712-1769553711           /            NaN    NaN      NaN   
2  68452117.1769553712-1769553711  /product/2            NaN   18.5      USD   
3  68452117.1769553712-1769553711  /product/2            NaN   18.5      USD   
4  68452117.1769553712-1769553711           /            NaN    NaN      NaN   

   item_id   item_name  price  quantity  
0      NaN         NaN    NaN       NaN  
1      2.0   Product 2   18.5       1.0  
2      2.0   Product 2   18.5       1.0  
3      2.0   Product 2   18.5       1.0  
4     16.0  Product 16   14.0       1.0

# Data Cleaning

Preparing `orders.csv`, `purchases.csv`, and `sessions.csv` for product analysis.

## Standardize column names

Ensure column names are made up of all lowercase letters, have no trailing or leading whitespace, and internal spaces are separated by one underscore.

In [12]:
df_orders.info() # orders.csv

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   transaction  id  60 non-null     object 
 1    session_key     60 non-null     object 
 2   purchase_ts      60 non-null     object 
 3   currency         60 non-null     object 
 4   revenue          60 non-null     float64
 5   units            60 non-null     int64  
 6   item_lines       60 non-null     int64  
 7   distinct_items   60 non-null     int64  
dtypes: float64(1), int64(3), object(4)
memory usage: 3.9+ KB


In [13]:
# 1) remove leading/trailing spaces
# 2) collapse whitespace to an underscore
df_orders.columns = (df_orders.columns.str.strip().str.replace(r"\s+", "_", regex=True))

In [14]:
df_products.info() # purchases.csv

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ts              67 non-null     object 
 1   user_pseudo_id  67 non-null     float64
 2    session_key    67 non-null     object 
 3   transaction id  67 non-null     object 
 4   order_value     67 non-null     float64
 5   currency        67 non-null     object 
 6   item_id         67 non-null     int64  
 7   item_name       67 non-null     object 
 8   price           67 non-null     float64
 9   quantity        67 non-null     int64  
 10  item_revenue    67 non-null     float64
dtypes: float64(4), int64(2), object(5)
memory usage: 5.9+ KB


In [15]:
df_products.columns = (df_products.columns.str.strip().str.replace(r"\s+", "_", regex=True))

In [16]:
df_sessions.info() # sessions.csv

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 704 entries, 0 to 703
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   session_key               704 non-null    object 
 1   user_pseudo_id            704 non-null    float64
 2   session_start_ts          704 non-null    object 
 3   landing_page_path         698 non-null    object 
 4   session_viewport_bucket   698 non-null    object 
 5   session_viewport_profile  233 non-null    object 
 6   session_viewport_group    233 non-null    object 
 7   country                   704 non-null    object 
 8   region                    704 non-null    object 
 9   device_category           704 non-null    object 
 10  os                        704 non-null    object 
 11  browser                   704 non-null    object 
dtypes: float64(1), object(11)
memory usage: 66.1+ KB


In [17]:
df_checkout.info() # checkout_dropoff.csv

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 7 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   session_key               120 non-null    object 
 1   begin_checkout_ts         120 non-null    object 
 2   purchase_ts               34 non-null     object 
 3   transaction_id            34 non-null     object 
 4   purchase_value            34 non-null     float64
 5   is_dropoff                120 non-null    int64  
 6   begin_checkout_items_len  120 non-null    int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 6.7+ KB


In [18]:
df_funnel.info() # events_funnel.csv

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6390 entries, 0 to 6389
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ts              6390 non-null   object 
 1   event_name      6390 non-null   object 
 2   user_pseudo_id  6390 non-null   float64
 3   session_key     6390 non-null   object 
 4   page_path       6311 non-null   object 
 5   transaction_id  67 non-null     object 
 6   value           4247 non-null   float64
 7   currency        4247 non-null   object 
 8   item_id         6123 non-null   float64
 9   item_name       6123 non-null   object 
 10  price           6123 non-null   float64
 11  quantity        6123 non-null   float64
dtypes: float64(5), object(7)
memory usage: 599.2+ KB


## Type enforcement

Convert IDs to strings, timestamps to datetimes, numeric metrics to numeric types, and categorical objects to categorical types to make grouping and plotting reliable.

Print what categories exist for each categorical column.

In [19]:
# orders.csv
ORDER_ID_COLS = {"transaction_id", "session_key"}
ORDER_DATE_COLS = {"purchase_ts"}
ORDER_FLOAT_COLS = {"revenue"}
ORDER_INT_COLS = {"units", "item_lines", "distinct_items"}
ORDER_CATEGORICAL_COLS = {"currency"}

# Cast in place
for c in df_orders.columns:
    if c in ORDER_ID_COLS:
        df_orders[c] = df_orders[c].astype("string")
    elif c in ORDER_DATE_COLS:
        df_orders[c] = pd.to_datetime(df_orders[c], errors="coerce")
    elif c in ORDER_FLOAT_COLS:
        df_orders[c] = pd.to_numeric(df_orders[c], errors="coerce").astype("Float64")
    elif c in ORDER_INT_COLS:
        df_orders[c] = pd.to_numeric(df_orders[c], errors="coerce").astype("Int64")
    elif c in ORDER_CATEGORICAL_COLS:
        df_orders[c] = df_orders[c].astype("category")

print("categories of the `currency` column:", list(df_orders["currency"].cat.categories))

categories of the `currency` column: ['USD']


In [20]:
# purchases.csv
ID_COLS = {"user_pseudo_id", "session_key", "transaction_id", "item_id"}
DATE_COLS = {"ts"}
FLOAT_COLS = {"order_value", "price", "item_revenue"}
INT_COLS = {"quantity"}
CATEGORICAL_COLS = {"currency"}
STRING_COLS = {"item_name"}

for c in df_products.columns:
    if c in ID_COLS:
        df_products[c] = df_products[c].astype("string")
    elif c in DATE_COLS:
        df_products[c] = pd.to_datetime(df_products[c], errors="coerce")
    elif c in FLOAT_COLS:
        df_products[c] = pd.to_numeric(df_products[c], errors="coerce").astype("Float64")
    elif c in INT_COLS:
        df_products[c] = pd.to_numeric(df_products[c], errors="coerce").astype("Int64")
    elif c in STRING_COLS:
        df_products[c] = df_products[c].astype("string")
    elif c in CATEGORICAL_COLS:
        df_products[c] = df_products[c].astype("category")

print("categories of the `currency` column:", list(df_products["currency"].cat.categories))

categories of the `currency` column: ['USD']


In [21]:
# sessions.csv
ID_COLS = {"session_key", "user_pseudo_id"}
DATE_COLS = {"session_start_ts"}
STRING_COLS = {"landing_page_path"}
CATEGORICAL_COLS = {"session_viewport_bucket", "session_viewport_profile", "session_viewport_group", "country", "region", "device_category", "os", "browser"}

for c in df_sessions.columns:
    if c in ID_COLS:
        df_sessions[c] = df_sessions[c].astype("string")
    elif c in DATE_COLS:
        df_sessions[c] = pd.to_datetime(df_sessions[c], errors="coerce", utc=True)
    elif c in STRING_COLS:
        df_sessions[c] = df_sessions[c].astype("string")
    elif c in CATEGORICAL_COLS:
        df_sessions[c] = df_sessions[c].astype("category")


for c in (CATEGORICAL_COLS):
    if str(df_sessions[c].dtype) == "category":
        cats = list(df_sessions[c].cat.categories)
        print(f"categories of `{c}`: {cats}")

categories of `session_viewport_group`: ['desktop', 'mobile', 'tablet']
categories of `session_viewport_profile`: ['1280x720', '1280x800', '1366x768', '1488x840', '1536x864', '1920x1080', '360x800', '390x844', '393x873', '412x915', '768x1024', '800x1280', '820x1180']
categories of `os`: ['Macintosh', 'iOS']
categories of `device_category`: ['desktop', 'mobile']
categories of `region`: ['Ontario']
categories of `browser`: ['Chrome', 'Safari']
categories of `country`: ['Canada']
categories of `session_viewport_bucket`: ['360-389', '390-429', '430-767', '<360', '>=768']


In [22]:
# checkout_dropoff.csv
CHECKOUT_ID_COLS = {"session_key", "transaction_id"}
CHECKOUT_DATE_COLS = {"begin_checkout_ts", "purchase_ts"}
CHECKOUT_FLOAT_COLS = {"purchase_value"}
CHECKOUT_INT_COLS = {"is_dropoff", "begin_checkout_items_len"}

for c in df_checkout.columns:
    if c in CHECKOUT_ID_COLS:
        df_checkout[c] = df_checkout[c].astype("string")
    elif c in CHECKOUT_DATE_COLS:
        df_checkout[c] = pd.to_datetime(df_checkout[c], errors="coerce")
    elif c in CHECKOUT_FLOAT_COLS:
        df_checkout[c] = pd.to_numeric(df_checkout[c], errors="coerce").astype("Float64")
    elif c in CHECKOUT_INT_COLS:
        df_checkout[c] = pd.to_numeric(df_checkout[c], errors="coerce").astype("Int64")

In [23]:
# events_funnel.csv
FUNNEL_ID_COLS = {"user_pseudo_id", "session_key", "transaction_id", "item_id"}
FUNNEL_DATE_COLS = {"ts"}
FUNNEL_FLOAT_COLS = {"value", "price"}
FUNNEL_INT_COLS = {"quantity"}
FUNNEL_CATEGORICAL_COLS = {"event_name", "currency"}
FUNNEL_STRING_COLS = {"page_path", "item_name"}

for c in df_funnel.columns:
    if c in FUNNEL_ID_COLS:
        df_funnel[c] = df_funnel[c].astype("string")
    elif c in FUNNEL_DATE_COLS:
        df_funnel[c] = pd.to_datetime(df_funnel[c], errors="coerce")
    elif c in FUNNEL_FLOAT_COLS:
        df_funnel[c] = pd.to_numeric(df_funnel[c], errors="coerce").astype("Float64")
    elif c in FUNNEL_INT_COLS:
        df_funnel[c] = pd.to_numeric(df_funnel[c], errors="coerce").astype("Int64")
    elif c in FUNNEL_STRING_COLS:
        df_funnel[c] = df_funnel[c].astype("string")
    elif c in FUNNEL_CATEGORICAL_COLS:
        df_funnel[c] = df_funnel[c].astype("category")

for c in FUNNEL_CATEGORICAL_COLS:
    if c in df_funnel.columns and str(df_funnel[c].dtype) == "category":
        cats = list(df_funnel[c].cat.categories)
        print(f"categories of `{c}`: {cats}")

categories of `currency`: ['USD']
categories of `event_name`: ['add_to_cart', 'begin_checkout', 'purchase', 'select_item', 'view_item']


## Handling Missing Values

Identify columns with missing values and decide whether to drop, impute, or leave as missing to improve data quality.

`missing_report` is a helper function to identify columns with missing values and calculate what percentage of the column contains missing values.

In [24]:
def missing_report(df, name):

    # for each column, find amount of missing values
    missing_val_count_per_col = df.isna().sum()
    
    row_count = df.shape[0] # row count

    missing_percentage = (missing_val_count_per_col/row_count) * 100 # percentage the missing counts

    missing_percentage = missing_percentage[missing_percentage > 0] # filter out complete rows

    missing_col_count = len(missing_percentage) # count how many cols have missing values

    if missing_col_count== 0:
        print(f"`{name}.csv` has 0 columns with missing values.\n")
        return

    missing_percentage = missing_percentage.sort_values(ascending=False) # sort by highest missing percentage

    print(f"`{name}.csv` has {missing_col_count} columns with missing values:")

    missing_df = missing_percentage.reset_index()
    missing_df.columns = ["column", "missing_%"]
    display(missing_df)


missing_report(df_orders, "orders")
missing_report(df_products, "purchases")
missing_report(df_sessions, "sessions")
missing_report(df_checkout, "checkout_dropoff")
missing_report(df_funnel, "events_funnel")

`orders.csv` has 0 columns with missing values.

`purchases.csv` has 0 columns with missing values.

`sessions.csv` has 4 columns with missing values:


column  missing_%
0  session_viewport_profile  66.903409
1    session_viewport_group  66.903409
2         landing_page_path   0.852273
3   session_viewport_bucket   0.852273

`checkout_dropoff.csv` has 3 columns with missing values:


column  missing_%
0     purchase_ts  71.666667
1  transaction_id  71.666667
2  purchase_value  71.666667

`events_funnel.csv` has 8 columns with missing values:


column  missing_%
0  transaction_id  98.951487
1           value  33.536776
2        currency  33.536776
3         item_id   4.178404
4       item_name   4.178404
5           price   4.178404
6        quantity   4.178404
7       page_path   1.236307

In [25]:
# for df_sessions, look at the rows with missing `landing_page_path` column values or missing `session_viewport_bucket` column values.
bad_rows = df_sessions[df_sessions[["session_viewport_bucket", "landing_page_path"]].isna().any(axis=1)]
bad_rows

session_key      user_pseudo_id  \
0  1766310541.1769539027-1769539027   1766310541.176954   
1  1675157078.1769508380-1769508380  1675157078.1769507   
2   954119712.1769507864-1769507864   954119712.1769508   
3  1124376317.1769540077-1769540076   1124376317.176954   
4  1309400224.1769507695-1769507695  1309400224.1769507   
5   133880033.1769516863-1769516863  133880033.17695168   

                  session_start_ts landing_page_path session_viewport_bucket  \
0 2026-01-27 18:37:07.703452+00:00              <NA>                     NaN   
1 2026-01-27 10:06:20.526923+00:00              <NA>                     NaN   
2 2026-01-27 09:57:44.338723+00:00              <NA>                     NaN   
3 2026-01-27 18:54:37.022376+00:00              <NA>                     NaN   
4 2026-01-27 09:54:55.567927+00:00              <NA>                     NaN   
5 2026-01-27 12:27:44.381649+00:00              <NA>                     NaN   

  session_viewport_profile session_viewport_group country   region  \
0                      NaN                    NaN  Canada  Ontario   
1                      NaN                    NaN  Canada  Ontario   
2                      NaN                    NaN  Canada  Ontario   
3                      NaN                    NaN  Canada  Ontario   
4                      NaN                    NaN  Canada  Ontario   
5                      NaN                    NaN  Canada  Ontario   

  device_category         os browser  
0         desktop  Macintosh  Chrome  
1          mobile        iOS  Safari  
2          mobile        iOS  Safari  
3         desktop  Macintosh  Chrome  
4          mobile        iOS  Safari  
5         desktop  Macintosh  Chrome

In [26]:
# drop the bad rows in df_sessions
df_sessions = df_sessions.dropna(subset=["session_viewport_bucket", "landing_page_path"])

## Check for duplicate keys

Verify join keys are usable and that natural keys are unique.

In [27]:
# detect duplicate keys
print("`orders.csv` duplicate `transaction_id` count:", df_orders["transaction_id"].duplicated().sum())

print("`products.csv` duplicate (transaction_id, item_id) count:", df_products[["transaction_id", "item_id"]].duplicated().sum())

print("`sessions.csv` duplicate `session_key` count:", df_sessions["session_key"].duplicated().sum())

print("`checkout_dropoff.csv` duplicate `session_key` count:", df_checkout["session_key"].duplicated().sum())

# for the funnel dataset, we want to check for duplicate user events by checking for the same session key, event, timestamp, and keys in one row.
duplicate_key_cols = ["session_key", "event_name", "ts", "transaction_id", "item_id"]

# filter out non-existing columns
existing_duplicate_key_cols = [c for c in duplicate_key_cols if c in df_funnel.columns]

if not existing_duplicate_key_cols:
    print("`events_funnel.csv` exact-duplicate event rows count: 0")
else:
    duplicate_rows_mask = df_funnel.duplicated(subset=existing_duplicate_key_cols, keep="first")
    duplicate_row_count = int(duplicate_rows_mask.sum())
    print(f"`events_funnel.csv` exact-duplicate event rows count: {duplicate_row_count}")



`orders.csv` duplicate `transaction_id` count: 0
`products.csv` duplicate (transaction_id, item_id) count: 0
`sessions.csv` duplicate `session_key` count: 0
`checkout_dropoff.csv` duplicate `session_key` count: 0
`events_funnel.csv` exact-duplicate event rows count: 1904


#### `events_funnel.csv` has unwanted duplicates so the code investigates the duplicate rows

In [28]:
# events_funnel.csv shows duplicate event rows, so explore whats going on.
if not existing_duplicate_key_cols:
    pass
else:

    # find all rows where duplication is going on
    duplicate_groups_mask = df_funnel.duplicated(subset=existing_duplicate_key_cols, keep=False)
    duplicated_groups_df = df_funnel.loc[duplicate_groups_mask].copy()

    # calculate the percentage of duplicate rows
    
    # the variable 'duplicate_row_count' has the amount of duplicate rows
    total_rows = len(df_funnel) # total rows
    unique_rows = total_rows - duplicate_row_count # unique rows = total rows - duplicate rows
    duplicate_pct = (duplicate_row_count / unique_rows) * 100
    print(f"duplicate row percentage (dup rows/unique rows) {duplicate_pct:.2f}%")

    # sort so identical keys are adjacent
    duplicated_groups_df = duplicated_groups_df.sort_values(existing_duplicate_key_cols)

    duplicate_examples_df = duplicated_groups_df.head(10)
    display(duplicate_examples_df)


duplicate row percentage (dup rows/unique rows) 42.44%


ts event_name      user_pseudo_id  \
5224 2026-01-27 12:22:25.650868+00:00  view_item  1002196564.1769516   
5225 2026-01-27 12:22:25.650868+00:00  view_item  1002196564.1769516   
5217 2026-01-27 12:22:31.349198+00:00  view_item  1002196564.1769516   
5218 2026-01-27 12:22:31.349198+00:00  view_item  1002196564.1769516   
5206 2026-01-27 12:22:40.645853+00:00  view_item  1002196564.1769516   
5207 2026-01-27 12:22:40.645853+00:00  view_item  1002196564.1769516   
5194 2026-01-27 12:22:53.703996+00:00  view_item  1002196564.1769516   
5195 2026-01-27 12:22:53.703996+00:00  view_item  1002196564.1769516   
5177 2026-01-27 12:23:03.104222+00:00  view_item  1002196564.1769516   
5178 2026-01-27 12:23:03.104222+00:00  view_item  1002196564.1769516   

                           session_key    page_path transaction_id  value  \
5224  1002196564.1769516537-1769516536  /product/16           <NA>   14.0   
5225  1002196564.1769516537-1769516536  /product/16           <NA>   14.0   
5217  1002196564.1769516537-1769516536  /product/10           <NA>   22.0   
5218  1002196564.1769516537-1769516536  /product/10           <NA>   22.0   
5206  1002196564.1769516537-1769516536   /product/6           <NA>  15.25   
5207  1002196564.1769516537-1769516536   /product/6           <NA>  15.25   
5194  1002196564.1769516537-1769516536   /product/7           <NA>   21.0   
5195  1002196564.1769516537-1769516536   /product/7           <NA>   21.0   
5177  1002196564.1769516537-1769516536   /product/3           <NA>   24.0   
5178  1002196564.1769516537-1769516536   /product/3           <NA>   24.0   

     currency item_id   item_name  price  quantity  
5224      USD    16.0  Product 16   14.0         1  
5225      USD    16.0  Product 16   14.0         1  
5217      USD    10.0  Product 10   22.0         1  
5218      USD    10.0  Product 10   22.0         1  
5206      USD     6.0   Product 6  15.25         1  
5207      USD     6.0   Product 6  15.25         1  
5194      USD     7.0   Product 7   21.0         1  
5195      USD     7.0   Product 7   21.0         1  
5177      USD     3.0   Product 3   24.0         1  
5178      USD     3.0   Product 3   24.0         1

#### Drop non-unique duplicates

Currently, 1,904 rows are duplicates (42.44% of the dataset), likely from the backend pipeline. Drop duplicates and reset the index.”

In [29]:
# drop secondary duplicate rows so we only have unique rows
if not existing_duplicate_key_cols:
    pass
else:
    rows_before = len(df_funnel) # track original dataframe size (helps for the sanity check)

    secondary_duplicate_rows_mask = df_funnel.duplicated(subset=existing_duplicate_key_cols, keep="first") # mask of the secondary duplicate ows

    secondary_duplicates_count = int(secondary_duplicate_rows_mask.sum()) # count how many duplicate rows there are to report
    print(f"dropping {secondary_duplicates_count} duplicate rows from `events_funnel.csv`" )

    # keep only non-duplicate rows (remove duplicates)
    # drop repeats and reset index
    df_funnel = df_funnel.loc[~secondary_duplicate_rows_mask].reset_index(drop=True)

    rows_after = len(df_funnel)  # track size after
    print(f"rows: {rows_before} -> {rows_after}")

    # ensuring dropping duplicate rows went without problems
    remaining_secondary_duplicates = int(df_funnel.duplicated(subset=existing_duplicate_key_cols, keep="first").sum())  # check again
    print(f"`events_funnel.csv` exact-duplicate event rows count: {remaining_secondary_duplicates}")  # verify


dropping 1904 duplicate rows from `events_funnel.csv`
rows: 6390 -> 4486
`events_funnel.csv` exact-duplicate event rows count: 0


## Check for negatives

Ensure there are no negative values in columns that shouldn't have negative values.

In [30]:
def negative_report(df, name):
    """
    If there are numeric columns, check for negatives.
    If a df has no numeric columns, skip.
    Note: The current datasets shouldn't have any values that should be negative, so we can check all columns.
    """
    num_cols = df.select_dtypes(include=["number"]).columns
    if len(num_cols) == 0:
        return
    print(f"\n{name}")
    for c in num_cols:
        neg_count = int((df[c] < 0).sum())
        print(f"negatives in column `{c}`:", neg_count)

negative_report(df_orders, "orders.csv")
negative_report(df_products, "purchases.csv")
negative_report(df_sessions, "sessions.csv")
negative_report(df_checkout, "checkout_dropoff.csv")
negative_report(df_funnel, "events_funnel.csv")



orders.csv
negatives in column `revenue`: 0
negatives in column `units`: 0
negatives in column `item_lines`: 0
negatives in column `distinct_items`: 0

purchases.csv
negatives in column `order_value`: 0
negatives in column `price`: 0
negatives in column `quantity`: 0
negatives in column `item_revenue`: 0

checkout_dropoff.csv
negatives in column `purchase_value`: 0
negatives in column `is_dropoff`: 0
negatives in column `begin_checkout_items_len`: 0

events_funnel.csv
negatives in column `value`: 0
negatives in column `price`: 0
negatives in column `quantity`: 0


## Merge rows and Join coverage checks

Before EDA, verify that the merges used in `eda_part1.py` are safe.

All session keys in `df_sessions.session_key` were established to be unique in the 'Check for Duplicate Keys' section.

Therefore, we should check what percent of `df_orders` and `df_checkout`rows with a session_key exist in `df_sessions`. 100% is ideal.

In [31]:


# orders --> sessions coverage
orders_with_key = df_orders["session_key"].notna().sum()  # count of all rows in df_orders where session key is not null

orders_session_keys = df_orders.loc[df_orders["session_key"].notna(), "session_key"]  # all non-null session_key values in orders

orders_keys_exist_in_sessions = orders_session_keys.isin(df_sessions["session_key"])  # keep the rows where the session key exists in df_sessions

orders_linked = orders_keys_exist_in_sessions.sum()  # count of all rows that can be merged

if orders_with_key == orders_linked:
    print(f"every session key in `df_orders` can be found in `df_sessions`. 100% coverage")
else:
    merge_percentage = (orders_linked / orders_with_key) * 100
    print(f"not every session key in `df_orders` can be found in `df_sessions`. {merge_percentage:.2f}% coverage.")



# checkout --> sessions coverage
checkout_with_key = df_checkout["session_key"].notna().sum()  # count of all rows in df_checkout where session key is not null

checkout_session_keys = df_checkout.loc[df_checkout["session_key"].notna(), "session_key"]  # all non-null session_key values in checkout

checkout_keys_exist_in_sessions = checkout_session_keys.isin(df_sessions["session_key"])  # keep the rows where the session key exists in df_sessions

checkout_linked = checkout_keys_exist_in_sessions.sum()  # count of all rows that can be merged

if checkout_with_key == checkout_linked:
    print(f"every session key in `df_checkout` can be found in `df_sessions`. 100% coverage")
else:
    merge_percentage = (checkout_linked / checkout_with_key) * 100 if checkout_with_key else 0.0
    print(f"not every session key in `df_checkout` can be found in `df_sessions`. {merge_percentage:.2f}% coverage.")


every session key in `df_orders` can be found in `df_sessions`. 100% coverage
every session key in `df_checkout` can be found in `df_sessions`. 100% coverage


#### Combine to enrich datasets
- `orders_enriched`: left-join `df_orders` with `df_sessions` on `session_key`
- `checkout_enriched`: left-join `df_checkout` with `df_sessions` on `session_key`

In [32]:
orders_enriched_preview = df_orders.merge(df_sessions, on="session_key", how="left")  # left join (keeping all orders)
display(orders_enriched_preview.head(5))

transaction_id                       session_key  \
0  3c393f91-9dac-4d1b-bc9c-df7aaa77ae6c  1289644372.1769538534-1769538533   
1  8c216563-36e5-48a8-8d0e-27d675e01d08  1833988998.1768880128-1769147505   
2  65f1359a-5175-4722-b4f5-4a7e9d678b42   531065886.1769545136-1769545135   
3  d59d8c94-1761-4cd2-8628-666c7d2024cf  1833988998.1768880128-1769147505   
4  8ff144ec-7231-414a-909e-94452fdfc7c7  1712912575.1769517746-1769543575   

                       purchase_ts currency  revenue  units  item_lines  \
0 2026-01-27 18:31:37.192183+00:00      USD     8.99      1           1   
1 2026-01-23 06:45:24.912877+00:00      USD     8.99      1           1   
2 2026-01-27 20:19:19.650985+00:00      USD     8.99      1           1   
3 2026-01-23 06:31:45.199547+00:00      USD     9.75      1           1   
4 2026-01-27 20:00:04.168122+00:00      USD     9.75      1           1   

   distinct_items      user_pseudo_id                 session_start_ts  \
0               1  1289644372.1769538 2026-01-27 18:28:53.891992+00:00   
1               1   1833988998.176888 2026-01-23 05:51:45.701439+00:00   
2               1   531065886.1769545 2026-01-27 20:18:55.982030+00:00   
3               1   1833988998.176888 2026-01-23 05:51:45.701439+00:00   
4               1  1712912575.1769516 2026-01-27 19:52:55.475360+00:00   

  landing_page_path session_viewport_bucket session_viewport_profile  \
0                 /                   >=768                 1280x800   
1                 /                   >=768                      NaN   
2                 /                   >=768                 1280x800   
3                 /                   >=768                      NaN   
4                 /                   >=768                1920x1080   

  session_viewport_group country   region device_category         os browser  
0                 tablet  Canada  Ontario         desktop  Macintosh  Chrome  
1                    NaN  Canada  Ontario         desktop  Macintosh  Chrome  
2                 tablet  Canada  Ontario         desktop  Macintosh  Chrome  
3                    NaN  Canada  Ontario         desktop  Macintosh  Chrome  
4                desktop  Canada  Ontario         desktop  Macintosh  Chrome

In [33]:
checkout_enriched_preview = df_checkout.merge(df_sessions, on="session_key", how="left")  # left join keeps all checkout rows
display(checkout_enriched_preview.head(5))

session_key                begin_checkout_ts  \
0    68452117.1769553712-1769553711 2026-01-27 22:44:20.992607+00:00   
1   591413466.1769553485-1769553484 2026-01-27 22:38:24.240412+00:00   
2   618581068.1769553392-1769553391 2026-01-27 22:36:37.685767+00:00   
3    43482501.1769517775-1769552743 2026-01-27 22:26:50.381091+00:00   
4  2013144899.1769517745-1769551745 2026-01-27 22:10:44.128115+00:00   

                       purchase_ts                        transaction_id  \
0                              NaT                                  <NA>   
1 2026-01-27 22:38:24.430573+00:00  4753c1b2-75ce-4c57-ada0-775eaa3fa757   
2                              NaT                                  <NA>   
3                              NaT                                  <NA>   
4 2026-01-27 22:10:44.174588+00:00  c51fca48-0ede-4edc-9692-538a3a6a980d   

   purchase_value  is_dropoff  begin_checkout_items_len     user_pseudo_id  \
0            <NA>           1                         0  68452117.17695537   
1            39.0           0                         0  591413466.1769553   
2            <NA>           1                         0  618581068.1769552   
3            <NA>           1                         0  43482501.17695177   
4            43.5           0                         0  2013144899.176952   

                  session_start_ts landing_page_path session_viewport_bucket  \
0 2026-01-27 22:41:51.980342+00:00                 /                 390-429   
1 2026-01-27 22:38:05.195331+00:00                 /                   >=768   
2 2026-01-27 22:36:32.238623+00:00                 /                   >=768   
3 2026-01-27 22:25:43.284771+00:00                 /                 390-429   
4 2026-01-27 22:09:05.398625+00:00                 /                   >=768   

  session_viewport_profile session_viewport_group country   region  \
0                  393x873                 mobile  Canada  Ontario   
1                1920x1080                desktop  Canada  Ontario   
2                 1280x800                 tablet  Canada  Ontario   
3                  412x915                 mobile  Canada  Ontario   
4                 1536x864                desktop  Canada  Ontario   

  device_category         os browser  
0         desktop  Macintosh  Chrome  
1         desktop  Macintosh  Chrome  
2         desktop  Macintosh  Chrome  
3         desktop  Macintosh  Chrome  
4         desktop  Macintosh  Chrome